### PR TITLE
Libraries BOM: upgrading google-cloud-core, google-cloud-bom, and google-cloud-storage

### PR DESCRIPTION
**Pull request by suztomo**
as per https://github.com/googleapis/java-cloud-bom/pull/1161#issuecomment-724897271

@stephaniewang526

### CODE REVIEW

**Comment by suztomo** [comment]
Failed

```
Require upper bound dependencies error for org.threeten:threetenbp:1.4.4 paths to dependency are:
+-com.google.cloud.tools.opensource:upper-bounds-check:2.0.0-SNAPSHOT
  +-com.google.cloud:google-cloud-storage:1.113.1
    +-org.threeten:threetenbp:1.4.4
and
+-com.google.cloud.tools.opensource:upper-bounds-check:2.0.0-SNAPSHOT
  +-com.google.cloud:google-cloud-core:1.93.10
    +-org.threeten:threetenbp:1.5.0
]
```

**Comment by suztomo** [comment]
The `google.cloud.java.version` is the highest 0.142.0 already.

**Comment by suztomo** [comment]
Changing the order (e01db08) does not help:

```
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for com.fasterxml.jackson.core:jackson-core:2.11.1 paths to dependency are:
+-com.google.cloud.tools.opensource:upper-bounds-check:2.0.0-SNAPSHOT
  +-com.google.cloud:google-cloud-storage:1.113.1
    +-com.fasterxml.jackson.core:jackson-core:2.11.1
and
+-com.google.cloud.tools.opensource:upper-bounds-check:2.0.0-SNAPSHOT
  +-com.google.cloud:google-cloud-core:1.93.10
    +-com.google.http-client:google-http-client-jackson2:1.38.0
      +-com.fasterxml.jackson.core:jackson-core:2.11.3
]

```

**Comment by elharo** [comment]
This looks like it requires a new shared dependencies that updates the http client to 1.38.0.

**Comment by stephaniewang526** [comment]
> This looks like it requires a new shared dependencies that updates the http client to 1.38.0.

it might make sense for libraries-bom to import shared-dependencies BOM.

**Comment by elharo** [comment]
I think we need to push a new google-cloud-core and update to that to resolve the http client issue.

**Comment by elharo** [comment]
I think flattening might be somehow pushing google-cloud-storage's version of com.fasterxml.jackson.core:jackson-core backwards from 2.11.3 to 2.11.1. In normal resolution it picks 2.11.3:

```
[INFO] com.google.cloud:google-cloud-storage:jar:1.113.4-SNAPSHOT
[INFO] \- com.google.http-client:google-http-client-jackson2:jar:1.37.0:compile
[INFO]    \- com.fasterxml.jackson.core:jackson-core:jar:2.11.3:compile
```

**Comment by stephaniewang526** [comment]
> I think flattening might be somehow pushing google-cloud-storage's version of com.fasterxml.jackson.core:jackson-core backwards from 2.11.3 to 2.11.1. In normal resolution it picks 2.11.3:
> 
> ```
> [INFO] com.google.cloud:google-cloud-storage:jar:1.113.4-SNAPSHOT
> [INFO] \- com.google.http-client:google-http-client-jackson2:jar:1.37.0:compile
> [INFO]    \- com.fasterxml.jackson.core:jackson-core:jar:2.11.3:compile
> ```

Isn't the above dependency tree printed out from the java-storage repo which is using the flattened pom.xml? Why would flattening be the problem then? As mentioned before, flattening should not change dependency resolution order.

@suztomo the dependency tree you provided above appear to have an older version of google-cloud-storage -- 1.113.1 instead of 1.113.3([the one we are interested in](https://github.com/googleapis/java-cloud-bom/pull/1161)) which is pulling in the [older version of `+-com.fasterxml.jackson.core:jackson-core:2.11.1`](https://search.maven.org/artifact/com.google.cloud/google-cloud-storage/1.113.1/jar).

**Comment by elharo** [comment]
The dependency tree I printed out is from head of java-storage. This reads the pom.xml in that repo, pre-flattening I think. 

Flattening can change dependency resolution since it pulls all dependencies up to the top level. Dependencies that might have been overridden before now have higher precedence. Not sure whether that's happening here, but we've certainly noticed it before.

**Comment by stephaniewang526** [comment]
> The dependency tree I printed out is from head of java-storage. This reads the pom.xml in that repo, pre-flattening I think.
> 
> Flattening can change dependency resolution since it pulls all dependencies up to the top level. Dependencies that might have been overridden before now have higher precedence. Not sure whether that's happening here, but we've certainly noticed it before.

Client libraries by default uses the flattened pom, not the original pom.

This is not the issues here as I commented above -- Tomo is looking at an older version of google-cloud-storage client.

**Comment by elharo** [comment]
The java-storage version the test sees is the most recent one in google-cloud-bom. That is, we get com.google.cloud:google-cloud-storage:1.113.1 from google-cloud-bom 0.142.0

**Comment by suztomo** [comment]
The circular dependency (not in the sense that we declare in pom.xml) I observe is

1. the google-cloud-bom wants to upgrade the `com.google.cloud:google-cloud-storage` to the latest version "1.113.3" (https://github.com/googleapis/java-cloud-bom/pull/1161), but this PR does not pass [the check](https://github.com/googleapis/java-cloud-bom/pull/1161#issuecomment-724897271), because the libraries BOM master does not have the latest google-cloud-core.
1. the libraries BOM master wants to upgrade the google-cloud-core to newer version (this PR #1798), but this PR does not pass [the check](https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1798#issuecomment-724955965), because the latest google-cloud-bom has `com.google.cloud:google-cloud-storage:1.113.1` (old)

**Comment by stephaniewang526** [comment]
> The circular dependency (not in the sense that we declare in pom.xml) I observe is
> 
> 1. the google-cloud-bom wants to upgrade the `com.google.cloud:google-cloud-storage` to the latest version "1.113.3" ([googleapis/java-cloud-bom#1161](https://github.com/googleapis/java-cloud-bom/pull/1161)), but this PR does not pass [the check](https://github.com/googleapis/java-cloud-bom/pull/1161#issuecomment-724897271), because the libraries BOM master does not have the latest google-cloud-core.
> 2. the libraries BOM master wants to upgrade the google-cloud-core to newer version (this PR #1798), but this PR does not pass [the check](https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/1798#issuecomment-724955965), because the latest google-cloud-bom has `com.google.cloud:google-cloud-storage:1.113.1` (old)

How do you suggest we fix this? Should we continue checking for bom-upperbound?

**Comment by stephaniewang526** [comment]
> The java-storage version the test sees is the most recent one in google-cloud-bom. That is, we get com.google.cloud:google-cloud-storage:1.113.1 from google-cloud-bom 0.142.0

I am unable to see the circleci here but should we be checking against the master branch of google-cloud-bom instead? Since one check is checking in the latest released version (the test here) and another is checking against master branch (the bom-upperbound check in google-cloud-bom).

**Comment by suztomo** [comment]
I think using the master branch does not solve this problem, because the master branch of google-cloud-bom has old google-cloud-storage until googleapis/java-cloud-bom#1161 is merged.

**Comment by suztomo** [comment]
> How do you suggest we fix this? 

- googleapis/java-cloud-bom#1161 to ignore the failing upper bound check.
- Release a new google-cloud-bom with the latest google-cloud-storage.
- the libraries BOM master upgrades both the google-cloud-bom and the latest google-cloud-core at once. (This should pass the checks)

File: boms/cloud-oss-bom/pom.xml [hunk 1]
```diff
@@ -237,11 +237,16 @@
...
-         <version>1.93.9</version>
+         <version>1.94.0</version>
          <type>pom</type>
          <scope>import</scope>
        </dependency>
-       <dependency>
+      <dependency>
+        <groupId>com.google.cloud</groupId>
+        <artifactId>google-cloud-storage</artifactId>
+        <version>1.113.3</version>
```
Comment (suztomo) [inline boms/cloud-oss-bom/pom.xml:247]: This solves the upper bound check failure for threetenbp.

**Comment by suztomo** [comment]
Once we merge #1800 , the build should pass.

**Review by elharo — APPROVED** [review]

**Pull request merged**